flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/catalog/DatabaseCalciteSchema.java

```diff
@@ -66,109 +68,154 @@
 class DatabaseCalciteSchema extends FlinkSchema {
 	private final String databaseName;
 	private final String catalogName;
-	private final Catalog catalog;
+	private final CatalogManager catalogManager;
 	// Flag that tells if the current planner should work in a batch or streaming mode.
 	private final boolean isStreamingMode;
 
-	public DatabaseCalciteSchema(String databaseName, String catalogName, Catalog catalog, boolean isStreamingMode) {
+	public DatabaseCalciteSchema(String databaseName, String catalogName, CatalogManager catalog, boolean isStreamingMode) {
 		this.databaseName = databaseName;
 		this.catalogName = catalogName;
-		this.catalog = catalog;
+		this.catalogManager = catalog;
 		this.isStreamingMode = isStreamingMode;
 	}
 
 	@Override
 	public Table getTable(String tableName) {
-		ObjectPath tablePath = new ObjectPath(databaseName, tableName);
+		ObjectIdentifier identifier = ObjectIdentifier.of(catalogName, databaseName, tableName);
+		return catalogManager.getTable(identifier)
+			.map(result -> {
+				CatalogBaseTable table = result.getTable();
+				if (result.isTemporary()) {
+					return convertTemporaryTable(new ObjectPath(databaseName, tableName), table);
+				} else {
+					return convertPermanentTable(
+						identifier.toObjectPath(),
+						table,
+						catalogManager.getCatalog(catalogName)
+							.flatMap(Catalog::getTableFactory)
+							.orElse(null)
+					);
+				}
+			})
+			.orElse(null);
+	}
 
-		try {
-			if (!catalog.tableExists(tablePath)) {
-				return null;
+	private Table convertPermanentTable(
+			ObjectPath tablePath,
+			CatalogBaseTable table,
+			@Nullable TableFactory tableFactory) {
```

Comment:
make it `Optional<TableFactory>`?

Reply:
Thought it is allowed in this context (as an argument to a private method) it is actually discouraged by the flink community(https://flink.apache.org/contributing/code-style-and-quality-java.html#java-optional). 

Usage of `Optional` here would require adding an annotation to suppress warnings.